### PR TITLE
Feature: function dxDrawCircle

### PR DESCRIPTION
**Pull request by CrosRoad95**
Client side function: `dxDrawCircle`
# Syntax:
```lua
bool dxDrawCircle( vector2 position, unsigned float radius [, float startAngle = 0, float stopAngle = 360, color theColor = white, color theCenterColor = theColor, int segments = 32, int ratio = 1, bPostGUI = false ])
```
# Arguments:
1. **position** - Center of circle.
2. **radius** - Radius of circle.
3. **startAngle** - From which angle circle become drawing.
4. **stopAngle** - On which angle circle stop drawing.
5. **theColor** - Color outside center.
6. **theCenterColor** - Color in center of circle.
7. **segments** - How many triangles you want to draw, higher = more smoothly circle.
8. **ratio** - Ration between width and height, 2 mean that circle is 2 times widther then higher.
9. **bPostGui** - Draw post gui?

# Examples:
1. Pacmans low segments example ( equal 5 ):
image
2. Function dxDrawRoundedRectangle based on dxDrawCircle:
image
3. Pacman ( White squere at center is dxDrawRectangle, for debugging if circle drawing properly. ):
image
4. Example how working `fRatio` argument:
https://gyazo.com/9c19a4441ef4cd2cc330d3c8c2083268
6. Example how working startAngle and stopAngle:
image
7. Example how working startAngle and stopAngle with segments = 5 :
image
8. Example of different theColor and theColorCenter
image
9. Default segments count:
```lua
dxDrawCircle(1500,150,20);
dxDrawCircle(1700,150,100);
dxDrawCircle(1500,250,50);
```
image

# The circle:
image

# Performance:
I put 200 pacmans with 64 segments then what couse ( on my medium pc ) about - 20fps.
Example script below which draw a lot of stuff doing small fps drop.

# Tips:
1. If you want to draw small circle, use low  amount of segments.
2. You can draw every pieces of circle individually to make colorful circle.

# Script:
[circle.zip](https://github.com/multitheftauto/mtasa-blue/files/2233522/circle.zip)

### CODE REVIEW

**Comment by Pirulax** [comment]
How does it looks like with the default resolution?

**Comment by CrosRoad95** [comment]
@Pirulax i added how looks circle with only first 3 arguments ( x, y and radius ).

**Comment by Citizen01** [comment]
Feels weird to have 0° on the right side of the circle instead of the top but I guess it's not coming from you and that's how the underlaying function works ?

File: Client/core/Graphics/CGraphics.cpp [hunk 1]
```diff
@@ -795,6 +795,105 @@ void CGraphics::DrawRectQueued(float fX, float fY, float fWidth, float fHeight,
...
+    vertLast.x = fX + fRadius * cos(stopAngle) * fRatio;
+    vertLast.y = fY + fRadius * sin(stopAngle) / fRatio;
+    vertLast.z = 0;
+    vertLast.color = color;
+    Points.push_back(vertLast);
+
+    if (Points.size() >= 3)
+    {
+        m_pDevice->SetTexture(0, 0);
+        m_pDevice->DrawPrimitiveUP(D3DPT_TRIANGLEFAN, Points.size() - 2, &Points[0], sizeof(stVertex));
```
Comment (Citizen01) [inline Client/core/Graphics/CGraphics.cpp]: Made some researches and it looks like we should avoid using `DrawPrimitiveUP` at all cost as it is slow.
Source: https://www.gamedev.net/forums/topic/99121-drawprimitiveup-vs-drawprimitive/

I didn't look what should be done to not use the UP version.

Maybe someone who already worked with that stuff could add his input ? I'm just putting a warning flag here.

Comment (ccw808) [reply Client/core/Graphics/CGraphics.cpp]: DrawPrimitive would be a better choice if the vertex data remained unchanged between frames. The dynamic nature of dxDraw functions means DrawPrimitiveUP is fine

**Comment by Pirulax** [comment]
I think the same about it starting from the right side, Im lazy to look at your code, but adding a -90 to the deg should solve ur problem.

**Comment by ccw808** [comment]
How does the performance compare to
https://wiki.multitheftauto.com/wiki/Shader_examples#dxDrawCircle

**Comment by CrosRoad95** [comment]
I tested dxDrawCircle from wiki, my function is 2-3 times faster ( with default arguments, you can decrees amount of segments what speed up drawing )

**Comment by Einheit-101** [comment]
This is amazing. DX circles were a mayor performance hog, but dxDrawCircle (Shader variant) was already pretty good. If this here is even 2-3 times faster it would drain no performance at all. 
At this point i would like to thank everyone who is putting a lot of effort into MTA lately.

**Comment by Pirulax** [comment]
The time your resource prints in render is the time it takes to add those circles to the render queue.

File: Client/core/Graphics/CGraphics.cpp [hunk 2]
```diff
@@ -795,6 +795,105 @@ void CGraphics::DrawRectQueued(float fX, float fY, float fWidth, float fHeight,
     AddQueueItem(Item, bPostGUI);
 }
 
+void CGraphics::DrawCircleQueued(float fX, float fY, float fRadius, float startAngle, float stopAngle, unsigned long color, unsigned long colorCenter, ushort fSegments, float fRatio, bool bPostGUI, bool bSubPixelPositioning)
```
Comment (ghost) [inline Client/core/Graphics/CGraphics.cpp]: Please use hungarian notation for consistency ```float startAngle, float stopAngle, unsigned long color, unsigned long colorCenter```.

File: Client/core/Graphics/CGraphics.cpp [hunk 3]
```diff
@@ -795,6 +795,105 @@ void CGraphics::DrawRectQueued(float fX, float fY, float fWidth, float fHeight,
...
+}
+
+
+struct stVertex
+{
+    float x, y, z;
+    D3DCOLOR color;
+};
+
+void CGraphics::DrawCircleInternal(float fX, float fY, float fRadius, float startAngle, float stopAngle, unsigned long color, unsigned long colorCenter, ushort fSegments, float fRatio, bool bPostGUI, bool bSubPixelPositioning)
```
Comment (ghost) [inline Client/core/Graphics/CGraphics.cpp]: same thing here

File: Client/core/Graphics/CGraphics.h [hunk 4]
```diff
@@ -81,6 +81,7 @@ class CGraphics : public CGraphicsInterface, public CSingleton<CGraphics>
     void DrawLine3D(const CVector& vecBegin, const CVector& vecEnd, unsigned long ulColor, float fWidth = 1.0f);
     void DrawRectangle(float fX, float fY, float fWidth, float fHeight, unsigned long ulColor, bool bSubPixelPositioning = false);
     void DrawStringOutline(const RECT& rect, unsigned long ulColor, const wchar_t* szText, unsigned long ulFormat, LPD3DXFONT pDXFont);
+    void DrawCircleInternal(float fX, float fY, float fRadius, float startAngle, float stopAngle, unsigned long color, unsigned long colorCenter, ushort fSegments, float fRatio, bool bPostGUI, bool bSubPixelPositioning);
```
Comment (ghost) [inline Client/core/Graphics/CGraphics.h]: Make sure to use hungarian notation for all parameters.

File: Client/sdk/core/CGraphicsInterface.h [hunk 5]
```diff
@@ -79,6 +79,7 @@ class CGraphicsInterface
 
     virtual void DrawLine3D(const CVector& vecBegin, const CVector& vecEnd, unsigned long ulColor, float fWidth = 1.0f) = 0;
     virtual void DrawRectangle(float fX, float fY, float fWidth, float fHeight, unsigned long ulColor, bool bSubPixelPositioning = false) = 0;
+    virtual void DrawCircleQueued(float fX, float fY, float fRadius, float startAngle, float stopAngle, unsigned long color, unsigned long colorCenter, ushort fSegments, float fRatio, bool bPostGUI, bool bSubPixelPositioning) = 0;
```
Comment (ghost) [inline Client/sdk/core/CGraphicsInterface.h]: Make sure to use hungarian notation for all parameters.

File: Client/core/Graphics/CGraphics.h [hunk 6]
```diff
@@ -141,6 +142,8 @@ class CGraphics : public CGraphicsInterface, public CSingleton<CGraphics>
                           unsigned long ulFormat, ID3DXFont* pDXFont = NULL, bool bPostGUI = false, bool bColorCoded = false, bool bSubPixelPositioning = false,
                           float fRotation = 0, float fRotationCenterX = 0, float fRotationCenterY = 0);
 
+    void DrawCircleQueued(float fX, float fY, float fRadius, float startAngle, float stopAngle, unsigned long color, unsigned long colorCenter, ushort fSegments, float fRatio, bool bPostGUI, bool bSubPixelPositioning = false);
```
Comment (ghost) [inline Client/core/Graphics/CGraphics.h]: Make sure to use hungarian notation for all parameters.

File: Client/core/Graphics/CGraphics.cpp [hunk 7]
```diff
@@ -795,6 +795,105 @@ void CGraphics::DrawRectQueued(float fX, float fY, float fWidth, float fHeight,
...
+    if (bSubPixelPositioning)
+    {
+        fX += 0.5f;
+        fY += 0.5f;
+    }
+
+    startAngle = D3DXToRadian(startAngle);
+    stopAngle = D3DXToRadian(stopAngle);
+
+    std::vector<stVertex> Points;
```
Comment (ghost) [inline Client/core/Graphics/CGraphics.cpp]: rename `Points` to `vecPoints`

File: Client/core/Graphics/CGraphics.cpp [hunk 8]
```diff
@@ -795,6 +795,105 @@ void CGraphics::DrawRectQueued(float fX, float fY, float fWidth, float fHeight,
...
+    
+    // first
+    stVertex vertFirst;
+    vertFirst.x = fX + fRadius * cos(startAngle) * fRatio;
+    vertFirst.y = fY + fRadius * sin(startAngle) / fRatio;
+    vertFirst.z = 0;
+    vertFirst.color = color;
+    Points.push_back(vertFirst);
+
+    bool out = false;
```
Comment (ghost) [inline Client/core/Graphics/CGraphics.cpp]: if you are not using `out` variable, can this be removed?

File: Client/mods/deathmatch/logic/luadefs/CLuaDrawingDefs.cpp [hunk 9]
```diff
@@ -402,6 +403,67 @@ int CLuaDrawingDefs::DxDrawRectangle(lua_State* luaVM)
...
 }
 
+int CLuaDrawingDefs::DxDrawCircle(lua_State* luaVM)
+{
+    CVector2D vecPosition;
+    float     radius;
+    float     startAngle;
+    float     stopAngle;
+    SColor    color;
+    SColor    colorCenter;
```
Comment (ghost) [inline Client/mods/deathmatch/logic/luadefs/CLuaDrawingDefs.cpp]: rename `radius` to `fRadius`

**Review by ghost — COMMENTED** [review]
I tested these functions with the script (circle.zip) provided. I was also able to draw around 900 circles and rectangles with full FPS, so it's indeed good performance-wise. If variables are renamed properly, this PR is ready to merge.

File: Client/mods/deathmatch/logic/luadefs/CLuaDrawingDefs.cpp [hunk 10]
```diff
@@ -402,6 +403,67 @@ int CLuaDrawingDefs::DxDrawRectangle(lua_State* luaVM)
...
+    argStream.ReadVector2D(vecPosition);
+    argStream.ReadNumber(fRadius);
+    argStream.ReadNumber(fStartAngle, 0);
+    argStream.ReadNumber(fStopAngle, 360);
+    argStream.ReadColor(color, 0xFFFFFFFF);
+    argStream.ReadColor(colorCenter, color);
+    argStream.ReadNumber(fSegments, 32);
+    argStream.ReadNumber(fRatio, 1);
+    argStream.ReadBool(bPostGUI, false);
+    argStream.ReadBool(bSubPixelPositioning, false);
```
Comment (ccw808) [inline Client/mods/deathmatch/logic/luadefs/CLuaDrawingDefs.cpp]: bSubPixelPositioning is for backwards compatibility with old dxDraw functions (which used integer positions). It is not required for new dxDraw functions

**Pull request merged**